### ISSUE TITLE
Support the `license-file` field for custom licenses

### ISSUE THREAD

**Issue by fschutt** [volks73/cargo-wix#49]
Currently, `cargo-wix` will ignore the `license-file` field and only work when the license given is a builtin one.

Ex. if I have a cargo project with `license-file="LICENSE"`, cargo-wix ignores it completeley and tells me that it couldn't generate an appropriate license file:

https://github.com/volks73/cargo-wix/blob/f4ad20ece9020368578b980b73878b9cd1012521/src/wix.rs#L263-L275

- You are only checking the `license` field here, not the `license-file` field. So it will fail to build the MSI if you are not specifying a license identifier, but a license in a seperate file.

**Comment by volks73** [comment]
The `license-file` field is not ignored, but it is used for the "[sidecar](https://en.wikipedia.org/wiki/Sidecar_file)" license file, not the EULA. See the [Concepts](https://volks73.github.io/cargo-wix/cargo_wix/index.html#concepts) section of the documentation:

>In addition to the WXS template, there are several license templates which are used to generate an End User License Agreement (EULA) during the cargo wix --init command. Depending on the license ID(s) in the license field for a package's manifest (Cargo.toml), a license file in the Rich Text Format (RTF) is generated from a template and placed in the wix folder. This RTF file is then displayed in the license dialog of the installer. See the help information on the --print-template option for information about supported licenses. If the license field is not used or the license ID is not supported, then the EULA will not be automatically created during initialization and it will have to be created manually with a text editor or other authoring tool.
>
>During creation of the installer, the cargo wix subcommand will look for a LICENSE file in the root of the project. If the file exists, then it is used as the source file for the License.txt file that is placed alongside the bin folder during installation of the project with the Windows installer (msi). If a LICENSE file does not exist, then the -l,--license option should be used to specify a path to a file that can be used.
>
>If a custom license is used, then the license-file field should be used with the package's manifest. The path specified for the license-file field is used to create a "sidecar" license file that is installed alongside the bin folder during installation. Basically, if a custom license is used, everything should be manually setup and modified.

Just a little bit further down in the code from the snippet you provided is:

```rust
if manifest.get("package").and_then(|p| p.as_table())
    .and_then(|t| t.get("license-file")).is_none() {
    let license_file = Path::new(DEFAULT_LICENSE_FILE_NAME);
    if !license_file.exists() {
        warn!("A '{}' file does not exist in the project root. Please consider adding such a file to avoid errors during installer creation.", DEFAULT_LICENSE_FILE_NAME);
    }
}
```

Since it is possible for a user of this sub-command to create his or her own RTF EULA and sidecar license file, the messages are only warnings. Upon review I do notice the messages are slightly confusing. The warning message you are receiving needs to be changed to indicate the _EULA_ could not be generated, not the license file, i.e.

>Could not generate an appropriate license file in the Rich Text Format (RTF). Please manually create one to avoid errors during installer creation.

should be changed to

>Could not generate an appropriate EULA in the Rich Text Format (RTF) from the project's 'license' field. Please manually create one to avoid errors during installer creation.

I will admit there appears to be an overall lack of distinction between the EULA and the license file in the documentation, and the EULA/license file management is a bit of a mess, but this is because there are so many ways to define and format a license for a project, and the sub-command is trying to use a license in two ways: (1) as the EULA for the installer's license page and it needs to be in the RTF, which is not a common file format for licenses in Rust/Cargo projects, and (2) as a sidecar file that is installed with the binary, but the sidecar file does _not_ need to be in the RTF. 

Initially, I did a brief review of the most common ways licenses are defined for a project and optimized the sub-command usage from there. I reviewed a variety of projects on [crates.io](https://www.crates.io/), specifically other cargo sub-commands, and noticed most used the `license` field and a text file (with and/or without a file extension were equally common) in the root of the project. The naming of the license text files varied as well. The most common licenses were MIT, Apache-2.0, and GPL-v3.0, but none of the projects formatted their license files in RTF. So, I prioritized using the `license` field for EULA generation and the `license-file` field for the sidecar file. 

I included three RTF templates for the three most commonly usage licenses while ignoring any license-related files, such as (LICENSE-APACHE, LICENSE-MIT, LICENSE-GPLv3, etc.) in the project's root to "automagically" generate an EULA for a project. If other licenses became common for users of this sub-command, I would be happy to create and include RTF templates for those as well or welcome pull requests that include new templates. I used predefined templates because I was not about to build or include a full-featured RTF parser or generator within the sub-command. All versions of Windows have WordPad, which can be used to create and read RTF files, and the RTF format is kind of a mess, too. The creation and management of RTF license files was, and is, already kind of a distraction. Templates was the path of least resistance.

I am still not sure how best to handle EULAs, license files, and RTF requirements, but I am open to ideas or alternatives.

**Comment by JMurph2015** [comment]
I think it would be expected behavior to use the `license-file` field as the EULA, since after all it stands for "End User License Agreement" and it also almost universally ends up being the license file itself.

**Comment by volks73** [comment]
That was my original thought and intention, to have the EULA be the side car license file. However, the more I think about it and make some significant changes to the implementation (#54, #55, #57, #59, #67, #68, and #69), keeping the EULA and side car license file explicitly separate adds greater flexibility.

The EULA for the installer must be a Rich Text Format (.rtf) file, which for the most part is a Windows-specific file format. I have not encountered its usage on other systems/platforms and its usage on Windows also appears to be limited. But, the side car license file has no such requirement on format. It can be whatever format the developer wants. For a strictly Windows-based project, the side car license file can be RTF and then the EULA and license file are the same. But, if the project is to be deployed on different platforms, like macOS or Linux, then the side car license file should probably be just a Text (.txt) or HTML (.html) file, which would be universally accessible on all platforms but could not be used as the EULA for the Windows installer, without conversion. It may contain the same content as the EULA, but it would be a different file. 

A single RTF file could be used for a cross-platform project and the subcommand could convert it to a more universal format during build time, but that would require embedding a RTF-to-TXT or RTF-to-HTML converter (pandoc?) into this cargo subcommand. This feels a little bit of overkill and possibly more work than I can handle, but would be super cool. Explicitly separating the EULA (Windows only) from the license file (any platform) starts to make more sense to me given a cross-platform project and limitations.

I am in the process of adding a `--eula` and `--license` options for the CLI (#57) as part of a more polished and refined work flow with subcommands (#54), which would be part of a v0.1.0 release once I get some time and tests written. If the same file is to be used for both, then these two options can have a path to the same file:

```dos
C:\> cargo wix init --eula "C:\Eula.rtf" --license "C:\Eula.rft"
```

Note, these changes also add the ability to disable/skip the EULA screen for the installer but keep the side car license file if desired.

I agree having a single RTF file would be the majority use-case, so maybe the default behavior should be to use the file for the `--eula` option as the side car license file unless the `--license` option is used to override it or a `--no-license` flag is used (a developer may not want to include a side car license file).

**Comment by JMurph2015** [comment]
Thanks for the update!  If you want to go for more flexibility, that's great too!  I guess it would be useful to explain my conundrum in case it informs your implementation:

So I'm developing a proprietary tool that uses `cargo-wix` for Windows packaging.  Unfortunately that means that I need a EULA and license that is totally custom.  This has two notable effects:

1. When I was writing up documentation for how to do binary builds, I realized that I would have to have the `License.rtf` _inside_ the `/wix` folder and committed into the repo.  Now this has a distinct disadvantage because I would be unable to safely run `cargo wix --purge` in order to reinitialize the `main.wxs`, and it generally makes the project more fragile by relying on a very particular folder structure (compounded by the naming requirement for the EULA).  Working from the documentation, it took me the better part of two hours to figure out this scheme to make everything happy.
2. I can't put anything like "Proprietary" or "Unlicensed" in the `license` field or else the `cargo wix --init` will fail.  So I have to omit it, which other tools don't appreciate.  It'd be nice if there were some way to handle that more gracefully that'd be great, but it's more a papercut than the above.

Also I do appreciate the irony in complaining about a proprietary tools' needs to an open source project, but these sorts of things would make the library more useful!

**Comment by volks73** [comment]
>Also I do appreciate the irony in complaining about a proprietary tools' needs to an open source project, but these sorts of things would make the library more useful!

No worries. Funny enough, this open source project came from almost your exact situation. I was developing a proprietary tool in Rust and wanted to distribute it to internal users with a Windows Installer. I had a custom EULA and license file. I created a similar subcommand to automate the installer creation and include in the build process. Afterwards, I thought it would be useful for others. So, I repackaged the subcommand, _not_ the proprietary tool, and created the crate. Part of repackaging of the subcommand as an open source project was making it more applicable and trying to accommodate more use cases, with a focus on open source projects using it. Some things clearly did not get thought well during the repackaging.

>...it took me the better part of two hours to figure out this scheme to make everything happy.

I am sorry this became such a time sink. I have been in similar situations, and it is frustrating. I do have a couple of suggestions that might help you with your problem:

(1) Can you use the `license-file` field instead of the `license` field in your Cargo.toml file? From the [cargo reference](https://doc.rust-lang.org/cargo/reference/manifest.html):

>\# If a project is using a nonstandard license, then this key may be specified in
>\# lieu of the above key and must point to a file relative to this manifest
>\# (similar to the readme key).
>license-file = "..."

and then, while the [documentation](https://volks73.github.io/cargo-wix/cargo_wix/index.html) does not directly discuss the EULA, other than about creating one from the `license` field if it is an open source license, there is:

>If a custom license is used, then the license-file field should be used with the package's manifest. The path specified for the license-file field is used to create a "sidecar" license file that is installed alongside the bin folder during installation. Basically, if a custom license is used, everything should be manually setup and modified.

A RTF license can be used for the `license-file` field and it can exist anywhere. For v0.0.5, I cannot remember off the top of my head if the `license-file` field is used for the EULA in the installer as well, but if you use the `license-file` instead of the `license` field does an error occur? No error should occur and if one is occurring, then I would be interested in the error message and filing a separate issue/bug report for it.

(2) You can edit the `wix\main.wxs` with a text editor. It is just an XML file. After using the `cargo wix --init` command, you can open the `wix\main.wxs` file and change the paths to the EULA from `wix\License.rtf` to whatever is appropriate. See:

```xml
...
<WixVariable Id='WixUILicenseRtf' Value='wix\License.rtf'>
...
```

to change the EULA file. Then, see

```xml
<Component Id='License' Guid='*' Win64='$(var.Win64)'>
    <File
        Id='LicenseFile'
        Name='$(var.LicenseName)'
        DiskId='1'
        Source='$(var.LicenseSource)'
        KeyPath='yes'/>
</Component>
```

and change the values for the `Name` and `Source` attribute to whatever is appropriate. The files can reside anywhere, but you will have to make these changes each time you `--purge` and `--init`. Again, I apologize for the lack of elegance in this solution.

(3) Finally, you can always try the bleeding edge of the subcommand and use the `master` branch of the code. It has the new `--eula` and `--license` options for the `init` subcommand, along with a bunch of other (potential?) improvements. I am in the process of adding tests after a significant refactoring, but so far the functionality appears to be working. You can use this unreleased version with the following command:

```dos
C:\>cargo install --git https://github.com/volks73/cargo-wix --force
```

then use

```dos
C:\>cargo wix init --help
```

for more information about the `--eula` and `--license` options. Warning, I have not changed the version number in the manifest yet, so the git version will still appear as `v0.0.5` if that is a problem. The documentation would also be out-of-date, but the documentation needs work anyways.

If you do go this route, then I would be interested in your feedback for this new and hopefully soon-to-be released version.

Thanks for the feedback and I hope this helps.

**Comment by volks73** [comment]
I think this has mostly been addressed with the new release (v0.1.0). It should be much clearer how the subcommand works in relation to licenses and EULAs.
